src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Instrument.netcore.cs

```diff
@@ -0,0 +1,75 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// Instrument{T} is the base class from which all non-observable instruments will inherit from.
+    /// </summary>
+    /// <remarks>
+    /// This class supports only the following generic parameter types: <see cref="byte" />, <see cref="short" />, <see cref="int" />, <see cref="long" />, <see cref="float" />, <see cref="double" />, and <see cref="decimal" />
+    /// </remarks>
+    public abstract partial class Instrument<T> : Instrument where T : struct
+    {
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct TagsBag
+        {
+            internal KeyValuePair<string, object?> Tag1;
+            internal KeyValuePair<string, object?> Tag2;
+            internal KeyValuePair<string, object?> Tag3;
+        }
+
+        /// <summary>
+        /// Record the measurement by notifying all <see cref="MeterListener" /> objects which listeneing to this instrument.
+        /// </summary>
+        /// <param name="measurement">The measurement value.</param>
+        /// <param name="tag">A key-value pair tag associated with the measurement.</param>
+        [SkipLocalsInit]
+        protected void RecordMeasurement(T measurement, KeyValuePair<string, object?> tag)
+        {
+            TagsBag tags;
+            tags.Tag1 = tag;
+
+            RecordMeasurement(measurement, MemoryMarshal.CreateReadOnlySpan(ref tags.Tag1, 1));
+        }
+
+        /// <summary>
+        /// Record the measurement by notifying all <see cref="MeterListener" /> objects which listeneing to this instrument.
+        /// </summary>
+        /// <param name="measurement">The measurement value.</param>
+        /// <param name="tag1">A first key-value pair tag associated with the measurement.</param>
+        /// <param name="tag2">A second key-value pair tag associated with the measurement.</param>
+        [SkipLocalsInit]
```

Comment:
Why do you need SkipLocalsInit here?

Reply:
`SkipLocalsInit` is ignored for GC references. GC references are always zero initialized to avoid reporting random pointers during GC. `TagsBag` is all GC references, so I would expect it to make no difference here.

Reply:
I'll remove the attribute then.